lib/explorer/data_frame.ex

```diff
@@ -104,39 +104,136 @@ defmodule Explorer.DataFrame do
   @enforce_keys [:data, :groups]
   defstruct [:data, :groups]
 
+  @type column_name :: atom() | String.t()
+  @type column :: column_name() | non_neg_integer()
+  @type columns :: [column] | Range.t()
+  @type column_names :: [column_name]
+  @type column_pairs(other) :: [{column(), other}] | %{column() => other}
+
   @default_infer_schema_length 1000
 
+  # Guards and helpers for columns
+
+  defguard is_column(column) when is_binary(column) or is_atom(column) or is_integer(column)
+  defguard is_column_name(column) when is_binary(column) or is_atom(column)
+  defguard is_column_pairs(columns) when is_list(columns) or is_map(columns)
+
+  # Normalize a column name to string
+  defp to_column_name(column) when is_binary(column), do: column
+  defp to_column_name(column) when is_atom(column), do: Atom.to_string(column)
+
+  # Normalize pairs of `{column, value}` where value can be anything.
+  # The `column` is only validated if it's an integer. We check that the index is present. 
+  defp to_column_pairs(df, pairs), do: to_column_pairs(df, pairs, & &1)
+
+  # The function allows to change the `value` for each pair.
+  defp to_column_pairs(df, pairs, value_fun)
+       when is_column_pairs(pairs) and is_function(value_fun, 1) do
+    existing_columns = names(df)
+
+    pairs
+    |> Enum.map_reduce(nil, fn
+      {column, value}, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column =
+          case Map.fetch(map, column) do
+            {:ok, column} -> column
+            :error -> raise ArgumentError, "column does not exist at index #{column}"
+          end
+
+        {{existing_column, value_fun.(value)}, map}
+
+      {column, value}, maybe_map when is_atom(column) ->
+        column = Atom.to_string(column)
+
+        {{column, value_fun.(value)}, maybe_map}
+
+      {column, value}, maybe_map when is_binary(column) ->
+        {{column, value_fun.(value)}, maybe_map}
+    end)
+    |> then(fn {pairs, _} -> pairs end)
+  end
+
+  defp column_index_map(names),
+    do: for({name, idx} <- Enum.with_index(names), into: %{}, do: {idx, name})
+
+  # Normalize column names without verifying if columns exist.
+  defp to_column_names(names),
+    do: for(name <- Enum.to_list(names), do: to_column_name(name))
+
+  # Normalize column names and raise if column does not exist
+  defp to_existing_columns(df, columns) do
+    existing_columns = names(df)
+
+    columns
+    |> Enum.to_list()
+    |> Enum.map_reduce(nil, fn
+      column, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column =
+          case Map.fetch(map, column) do
+            {:ok, column} -> column
+            :error -> raise ArgumentError, "column does not exist at index #{column}"
+          end
```

Comment:
Encapsulate this in a function to remove the duplication with `to_column_pairs`. :)